Index: src/e/util/FileUtilities.java
===================================================================
--- src/e/util/FileUtilities.java	(revision 3043)
+++ src/e/util/FileUtilities.java	(working copy)
@@ -11,6 +11,8 @@
  * Provides utilities for dealing with files and paths.
  */
 public class FileUtilities {
+    private static final String LIBS_SYSTEM_PROPERTY = "org.jessies.libraryDirectories";
+
     /**
      * Returns a new File for the given filename, coping with "~/".
      * Try not to ever use "new File(String)": use this instead.
@@ -363,8 +365,22 @@
     }
 
     public static void loadNativeLibrary(String libraryName) {
+        String directories = System.getProperty(LIBS_SYSTEM_PROPERTY);
+        if (directories == null) {
+          // Fallback to the standard approach, which is useful for integration with OSGI.
+          try {
+              System.loadLibrary(libraryName);
+              return;
+          }
+          catch (UnsatisfiedLinkError ex) {
+              UnsatisfiedLinkError explain = new UnsatisfiedLinkError("Failed to load "
+                + libraryName
+                + ".  Consider setting " + LIBS_SYSTEM_PROPERTY + " system property.");
+              explain.initCause(ex);
+              throw explain;
+          }
+        }
         String fileName = System.mapLibraryName(libraryName);
-        String directories = System.getProperty("org.jessies.libraryDirectories");
         for (String directory : directories.split(File.pathSeparator)) {
             File candidatePath = new File(directory, fileName);
             try {
