Index: native/all/pty/terminator_terminal_PtyProcess.cpp
===================================================================
--- native/all/pty/terminator_terminal_PtyProcess.cpp	(revision 1504)
+++ native/all/pty/terminator_terminal_PtyProcess.cpp	(working copy)
@@ -349,3 +349,9 @@
     listProcessesUsingTty(processNames, ttyFilename);
     return newStringUtf8(join(", ", processNames));
 }
+
+jlong terminator_terminal_PtyProcess::nativeGetPid() {
+    pid_t pid = processId.get();
+    return pid;
+}
+
Index: src/terminator/terminal/PtyProcess.java
===================================================================
--- src/terminator/terminal/PtyProcess.java	(revision 1504)
+++ src/terminator/terminal/PtyProcess.java	(working copy)
@@ -192,6 +192,10 @@
         return result;
     }
     
+    public long getPid() {
+      return nativeGetPid();
+    }
+    
     private native void nativeStartProcess(String executable, String[] argv, String workingDirectory) throws IOException;
     private native void nativeWaitFor() throws IOException;
     public native void destroy() throws IOException;
@@ -202,4 +206,7 @@
     public native void sendResizeNotification(Dimension sizeInChars, Dimension sizeInPixels) throws IOException;
     
     private native String nativeListProcessesUsingTty() throws IOException;
+
+    private native long nativeGetPid();
+
 }
