Index: native/all/pty/terminator_terminal_PtyProcess.cpp
===================================================================
--- native/all/pty/terminator_terminal_PtyProcess.cpp	(revision 1504)
+++ native/all/pty/terminator_terminal_PtyProcess.cpp	(working copy)
@@ -349,3 +349,9 @@
     listProcessesUsingTty(processNames, ttyFilename);
     return newStringUtf8(join(", ", processNames));
 }
+
+jlong terminator_terminal_PtyProcess::nativeGetChildPid() {
+    pid_t pid = processId.get();
+    return pid;
+}
+
Index: src/terminator/terminal/PtyProcess.java
===================================================================
--- src/terminator/terminal/PtyProcess.java	(revision 1504)
+++ src/terminator/terminal/PtyProcess.java	(working copy)
@@ -1,11 +1,19 @@
 package terminator.terminal;
 
-import e.util.*;
 import java.awt.Dimension;
-import java.io.*;
-import java.util.concurrent.*;
-import java.util.regex.*;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
+import e.util.FileUtilities;
+import e.util.Log;
+import e.util.ThreadUtilities;
+
 public class PtyProcess {
     private class PtyInputStream extends InputStream {
         /**
@@ -192,6 +200,10 @@
         return result;
     }
     
+    public long getChildPid() {
+      return nativeGetChildPid();
+    }
+    
     private native void nativeStartProcess(String executable, String[] argv, String workingDirectory) throws IOException;
     private native void nativeWaitFor() throws IOException;
     public native void destroy() throws IOException;
@@ -202,4 +214,7 @@
     public native void sendResizeNotification(Dimension sizeInChars, Dimension sizeInPixels) throws IOException;
     
     private native String nativeListProcessesUsingTty() throws IOException;
+
+    private native long nativeGetChildPid();
+
 }
